File: InterviewPrep.Core/Graphs/AdjacencyListGraph.cs
using System;
using System.Linq;
using System.Collections.Generic;

namespace InterviewPrep.Core.Graphs
{
    public class AdjacencyListGraph : IGraph
    {
        private readonly IDictionary<string, Vertex> _adjacencyDictionary;

        public Vertex this[string index] => _adjacencyDictionary[index];

        public AdjacencyListGraph(IDictionary<string, Vertex> adjacencyDictionary)
        {
            _adjacencyDictionary = adjacencyDictionary;
        }

        public IEnumerable<string> FindAdjacentNodes(string node)
        {
            var vertex = _adjacencyDictionary[node];
            var adjacentNodes = vertex.Neighbors;
            return adjacentNodes.Select(v => v.Vertex.Name);
        }

        public bool AreNodesAdjacent(string nodeOne, string nodeTwo)
        {
            var vertexOne = _adjacencyDictionary[nodeOne];
            var vertexTwo = _adjacencyDictionary[nodeTwo];
            return vertexOne.IsAdjacent(vertexTwo);
        }

        public IEnumerable<string> FindShortestPath(string nodeOne, string nodeTwo)
        {
            var tracker = new Dictionary<string, FringeItem>();
            var fringe = new List<FringeItem>
            {
                new FringeItem {Vertex = _adjacencyDictionary[nodeOne] }
            };

            while (fringe.Any())
            {
                var minFringeItem = fringe.Min();
                fringe.Remove(minFringeItem);
                tracker.Add(minFringeItem.Vertex.Name, minFringeItem);

                foreach (var neighbor in minFringeItem.Vertex.Neighbors)
                {
                    if (tracker.ContainsKey(neighbor.Vertex.Name)) continue;
                    var neighborFringeItem = fringe.FirstOrDefault(f => f.Vertex == neighbor.Vertex);
                    var newDistance = minFringeItem.Distance + neighbor.Weight;
                    if (neighborFringeItem == null)
                    {
                        fringe.Add(new FringeItem
                        {
                            Vertex = neighbor.Vertex,
                            LastItem = minFringeItem,
                            Distance = newDistance
                        });
                    }
                    else
                    {
                        if (newDistance >= neighborFringeItem.Distance) continue;
                        neighborFringeItem.Distance = newDistance;
                        neighborFringeItem.LastItem = minFringeItem;
                    }
                }
            }

            var temp = tracker[nodeTwo].FindPath().Select(f => f.Vertex.Name).Reverse();
            return temp;
        }
    }

    public class Vertex
    {
        public string Name { get; set; }
        public List<Neighbor> Neighbors { get; set; } = new List<Neighbor>();

        public bool IsAdjacent(Vertex other) => Neighbors.Any(n => n.Vertex == other);
    }

    public class Neighbor
    {
        public Vertex Vertex { get; set; }
        public int Weight { get; set; }
    }

    public class FringeItem : IComparable<FringeItem>
    {
        public Vertex Vertex { get; set; }
        public int Distance { get; set; }
        public FringeItem LastItem { get; set; }

        public int CompareTo(FringeItem other) => Distance.CompareTo(other.Distance);

        public IList<FringeItem> FindPath()
        {
            var list = new List<FringeItem>();
            FindPath(this, list);
            return list;
        }

        public void FindPath(FringeItem current, IList<FringeItem> list)
        {
            if (current == null)
                return;
            list.Add(current);
            FindPath(current.LastItem, list);
        }
    }
}

File: InterviewPrep.Concepts.VisualGraphs.Web/wwwroot/app/adjacencyListGraph.ts
/// <reference path="./graphics/pathcolorfactory" />

class AdjacencyListGraph implements IGraph
{

    constructor(public adjacencyDictionary: any)
    {
    }

    public areAdjacent(firstNode: Vertex, secondNode: Vertex): boolean
    {
        return false;
    }

    public findShortestPath(vertexOne: Vertex, vertexTwo: Vertex): Vertex[]
    {
        var fringe: FringeItem[] = [];
        var tracker: any = {};
        fringe.push(new FringeItem(vertexOne, null, 0));
        //fringe.push(new FringeItem(vertexTwo, null, 500));
        //fringe.push(new FringeItem(new Vertex("z"), null, 700));
        while (fringe.length > 0)
        {
            fringe.sort((a, b) => a.distance < b.distance ? 1 : -1);
            var minFringeItem = fringe.pop();
            tracker[minFringeItem.Vertex.name] = minFringeItem;
            for (var neighbor of minFringeItem.Vertex.neighbors)
            {
                //has neighbor shortest path been found?
                if (tracker[neighbor.vertex.name] != null) continue;

                //is neighbor currently on fringe?
                var neighborFringeItem = fringe.firstOrDefault(fi => fi.Vertex == neighbor.vertex);
                //add
                var newDistance = minFringeItem.distance + neighbor.weight;
                if (neighborFringeItem == null)
                {
                    var newFringeItem = new FringeItem(neighbor.vertex, minFringeItem, newDistance);
                    fringe.push(newFringeItem);
                }
                //update
                else
                {
                    if (newDistance >= neighborFringeItem.distance) continue;
                    neighborFringeItem.distance = newDistance;
                    neighborFringeItem.LastItem = minFringeItem;
                }
            }
        }
        var temp: FringeItem;
        for (var nodeName in tracker)
        {
            var fringeItem = <FringeItem>tracker[nodeName];
            if (fringeItem.Vertex == vertexTwo)
            {
                temp = fringeItem;
                break;
            }
        }
        var path = temp.findPrevious().map(fi => fi.Vertex);
        return path;
    }

    public update(): void
    {
        for (var nodeName in this.adjacencyDictionary)
        {
            var node = <Vertex>this.adjacencyDictionary[nodeName]
            node.update();
        }
    }

    public draw(context: CanvasRenderingContext2D, number: number): void
    {
        for (var nodeName in this.adjacencyDictionary)
        {
            var node = <Vertex>this.adjacencyDictionary[nodeName]
            node.drawConnections(context, number);
        }
        for (var nodeName in this.adjacencyDictionary)
        {
            var node = <Vertex>this.adjacencyDictionary[nodeName]
            node.draw(context);
        }
    }

    public getFullNodes(): number
    {
        var sum = 0;
        for (var nodeName in this.adjacencyDictionary)
        {
            var node = <Vertex>this.adjacencyDictionary[nodeName]
            sum += node.progress
        }
        return sum - 1;
    }

    public deactivateAll(): void
    {
        for (var nodeName in this.adjacencyDictionary)
        {
            var node = <Vertex>this.adjacencyDictionary[nodeName]
            node.deactivateFull();
        }
    }
}